cmd/snap/cmd_run.go

```diff
@@ -489,22 +489,36 @@ func (x *cmdRun) straceOpts() (opts []string, raw bool, err error) {
 	return opts, raw, nil
 }
 
-// isSnapRefreshConflictDetected detects if snap refreshed was started while not
-// holding the inhibition hint file lock.
+// checkSnapRunInhibitionConflict detects if snap refreshed/removed was started
+// while not holding the inhibition hint file lock.
 //
 // For context, on snap first install, the inhibition hint lock file is not created
-// so we cannot hold it. It is created after the first refresh. This allows for a
-// window where we don't hold the lock before the tracking cgroup is created where
-// a snap refresh could start.
-func isSnapRefreshConflictDetected(app *snap.AppInfo, hintFlock *osutil.FileLock) bool {
-	if !features.RefreshAppAwareness.IsEnabled() || app.IsService() || hintFlock != nil {
+// so we cannot hold it. It is created after the first refresh/remove. This allows
+// for a window where we don't hold the lock before the tracking cgroup is created
+// where a snap refresh/removal could start.
+func checkSnapRunInhibitionConflict(app *snap.AppInfo) error {
+	// Remove hint check takes precedence because we want to exit
+	snapName := app.Snap.InstanceName()
+	hint, _, err := runinhibit.IsLocked(snapName)
+	if err != nil {
+		return err
+	}
+	if hint == runinhibit.HintInhibitedForRemove {
+		return fmt.Errorf(i18n.G("cannot run %q, snap is being removed"), snapName)
+	}
+
+	if !features.RefreshAppAwareness.IsEnabled() || app.IsService() {
 		// Skip check
-		return false
+		return nil
 	}
 
 	// We started without a hint lock file, if it exists now this means that a
-	// refresh was started.
-	return osutil.FileExists(runinhibit.HintFile(app.Snap.InstanceName()))
+	// refresh might have started and retry is needed.
+	if osutil.FileExists(runinhibit.HintFile(snapName)) {
```

Comment:
do we need this check now? or can we just check hint != "" here ?

Reply:
yes, because `hint == ""` could mean two things, that the hint file is absent or exists but the snap is not inhibited. We want to specifically detect the later because it means that a refresh could have happened.

Reply:
that's because if a refresh happend we would use the wrong revision at this point? if that's the case the comment needs to be expanded a bit to mention both scenarios: ongoing refresh, finished refresh, because otherwise somebody will wonder again the same

Reply:
It's for both scenarios:
- if a refresh is finished, wrong revision will be used
- if a refresh is ongoing, depending on where we are in a refresh starting might either fail or use the wrong revision

The reasoning here was, we started without a lockfile, then just before running it appeared so it is better to retry.

Reply:
comment added, thank you!